packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -0,0 +1,696 @@
+
+function splitOnDot(name) {
+  return _splitOnDotCache[name] || (
+    _splitOnDotCache[name] = name.split('.')
+  );
+}
+
+function extractPivotName(name) {
+  return _extractPivotNameCache[name] || (
+    _extractPivotNameCache[name] = splitOnDot(name)[0]
+  );
+}
+
+function retrieveFromCurrentState(key) {
+  return function() {
+    return get(this.currentState, key);
+  };
+}
+
+/*
+  `InternalModel` is the Model class that we use internally inside Ember Data to represent models.
+  Internal ED methods should only deal with `InternalModel` objects. It is a fast, plain Javascript class.
+
+  We expose `DS.Model` to application code, by materializing a `DS.Model` from `InternalModel` lazily, as
+  a performance optimization.
+
+  `InternalModel` should never be exposed to application code. At the boundaries of the system, in places
+  like `find`, `push`, etc. we convert between Models and InternalModels.
+
+  We need to make sure that the properties from `InternalModel` are correctly exposed/proxied on `Model`
+  if they are needed.
+*/
+
+var InternalModel = function(type, id, store, container, data) {
+  this.type = type;
+  this.id = id;
+  this.store = store;
+  this.container = container;
+  this._data = data || {};
+  this.modelName = type.modelName;
+  this.errors = null;
+  this.dataHasInitialized = false;
+  this._deferredTriggers = [];
+  this._data = {};
+  this._attributes = Ember.create(null);
+  this._inFlightAttributes = Ember.create(null);
+  this._relationships = {};
+  this.currentState = RootState.empty;
+  /*
+    implicit relationships are relationship which have not been declared but the inverse side exists on
+    another record somewhere
+    For example if there was
+    ```
+      App.Comment = DS.Model.extend({
+        name: DS.attr()
+      })
+    ```
+    but there is also
+    ```
+      App.Post = DS.Model.extend({
+        name: DS.attr(),
+        comments: DS.hasMany('comment')
+      })
+    ```
+
+    would have a implicit post relationship in order to be do things like remove ourselves from the post
+    when we are deleted
+  */
+  this._implicitRelationships = Ember.create(null);
+  var model = this;
+  //TODO Move into a getter for better perf
+  this.eachRelationship(function(key, descriptor) {
+    model._relationships[key] = createRelationshipFor(model, descriptor, model.store);
+  });
+};
+
+InternalModel.prototype = {
+  isEmpty: retrieveFromCurrentState('isEmpty'),
+  isLoading: retrieveFromCurrentState('isLoading'),
+  isLoaded: retrieveFromCurrentState('isLoaded'),
+  isDirty: retrieveFromCurrentState('isDirty'),
+  isSaving: retrieveFromCurrentState('isSaving'),
+  isDeleted: retrieveFromCurrentState('isDeleted'),
+  isNew: retrieveFromCurrentState('isNew'),
+  isValid: retrieveFromCurrentState('isValid'),
+  dirtyType: retrieveFromCurrentState('dirtyType'),
+
+  constructor: InternalModel,
+  materializeRecord: function() {
+    // lookupFactory should really return an object that creates
```

Comment:
should we guard + throw if a re-materialization occurs? That seems like an error we would like to fail fast on.